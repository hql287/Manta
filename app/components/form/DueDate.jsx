// Libraries
import React, { Component } from 'react';
import PropTypes from 'prop-types';

// Custom Components
import { Section } from '../shared/Section';

// React Dates
import { SingleDatePicker } from 'react-dates';
import moment from 'moment';

// Styles
import styled from 'styled-components';
const DueDateContent = styled.div`
  display: flex;
`;

// Animation
import _withFadeInAnimation from '../shared/hoc/_withFadeInAnimation';

// Component
export class DueDate extends Component {
  constructor(props) {
    super(props);
    this.state = { focused: false, addDay:0 };
    this.onFocusChange = this.onFocusChange.bind(this);
    this.onDateChange = this.onDateChange.bind(this);
    this.clearDate = this.clearDate.bind(this);
    this.change = this.change.bind(this);
  }

  shouldComponentUpdate(nextProps, nextState) {
    return this.state !== nextState || this.props.dueDate != nextProps.dueDate;
  }

  onFocusChange() {
    this.setState({ focused: !this.state.focused });
  }

  onDateChange(date) {
    const selectedDate = date === null ? null : moment(date).toObject();
    this.props.updateFieldData('dueDate', { selectedDate });
    this.setState({addDay: 0});
  }

  clearDate() {
    this.onDateChange(null);
  }

  change(event){
    let result = new Date();
    result = result.setDate(result.getDate() + parseFloat(event.target.value));
    const selectedDate = moment(result).toObject();
    this.onDateChange(selectedDate);
    this.setState({ addDay: event.target.value});
    this.props.updateFieldData('dueDate', this.state);
  }

  render() {

    const { t, dueDate } = this.props;
    const selectedDate = dueDate.selectedDate
      ? moment(dueDate.selectedDate)
      : null;
    const days =  [7,10,30,60,90];
    const daysList = days.map(day => <option value={day}>{t('form:fields:dueDate:addDayNet')} {day} {t('form:fields:dueDate:addDaysToCurrentDate')}</option>);
    return (
      <Section>
        <label className="itemLabel">{t('form:fields:dueDate:name')}</label>
        <DueDateContent>
          <select name="addDay" id="addDay" placeholder={t('form:fields:dueDate:placeHolder')} onChange={this.change} value={this.state.addDay} >
            <option value="0">{t('form:fields:dueDate:addSomeDaysToCurrentDate')}</option>
            {daysList}
          </select>

          <SingleDatePicker
            id="invoice-duedate"
            placeholder={t('form:fields:dueDate:placeHolder')}
            firstDayOfWeek={1}
            withFullScreenPortal
            displayFormat="DD/MM/YYYY"
            hideKeyboardShortcutsPanel
            date={selectedDate}
            focused={this.state.focused}
            onFocusChange={this.onFocusChange}
            onDateChange={newDate => this.onDateChange(newDate)}
          />
          {selectedDate !== null && (
            <a
              className="clearDateBtn active"
              href="#"
              onClick={this.clearDate}
            >
              <i className="ion-close-circled" />
            </a>
          )}
        </DueDateContent>
      </Section>
    );
  }
}

DueDate.propTypes = {
  dueDate: PropTypes.object.isRequired,
  t: PropTypes.func.isRequired,
  updateFieldData: PropTypes.func.isRequired,
};

// Export
export default _withFadeInAnimation(DueDate);
