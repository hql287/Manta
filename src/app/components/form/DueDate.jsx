// Libraries
import React, { Component } from 'react';
import PropTypes from 'prop-types';

// Custom Components
import { Section } from '../shared/Section';
import DueDatePicker from './DueDatePicker';
import DueDateTerms from './DueDateTerms';

// Animation
import _withFadeInAnimation from '../shared/hoc/_withFadeInAnimation';

// Component
export class DueDate extends Component {
  constructor(props) {
    super(props);
    this.state = this.props.dueDate;
    this.toggleDatePicker = this.toggleDatePicker.bind(this);
    this.updateCustomDate = this.updateCustomDate.bind(this);
    this.updatePaymentTerm = this.updatePaymentTerm.bind(this);
    this.updateDueDate = this.updateDueDate.bind(this);
  }

  // Handle Clear Form
  componentWillReceiveProps(nextProps) {
    const { selectedDate, paymentTerm, useCustom } = nextProps.dueDate;
    if (selectedDate === null && paymentTerm === null && useCustom === true) {
      this.setState(nextProps.dueDate);
    }
  }

  toggleDatePicker() {
    this.setState({ useCustom: !this.state.useCustom }, () => {
      this.updateDueDate(this.state);
    });
  }

  updateCustomDate(selectedDate) {
    this.setState({ selectedDate }, () => {
      this.updateDueDate(this.state);
    });
  }

  updatePaymentTerm(paymentTerm) {
    this.setState({ paymentTerm }, () => {
      this.updateDueDate(this.state);
    });
  }

  updateDueDate(data) {
    this.props.updateFieldData('dueDate', data);
  }

  render() {
    const { t } = this.props;
    const { selectedDate, paymentTerm } = this.state;
    return (
      <Section>
        <label className="itemLabel">{t('form:fields:dueDate:name')}</label>
        {this.state.useCustom ? (
          <DueDatePicker
            t={t}
            selectedDate={selectedDate}
            updateCustomDate={this.updateCustomDate}
          />
        ) : (
          <DueDateTerms
            t={t}
            paymentTerm={paymentTerm}
            updatePaymentTerm={this.updatePaymentTerm}
          />
        )}
        <div>
          <div className="radio">
            <label>
              <input
                type="radio"
                onChange={this.toggleDatePicker}
                checked={this.state.useCustom === true}
                value="new"
              />
              Custom Date
            </label>
          </div>
          <div className="radio">
            <label>
              <input
                type="radio"
                onChange={this.toggleDatePicker}
                checked={this.state.useCustom === false}
                value="select"
              />
              Select Payment Term
            </label>
          </div>
        </div>
      </Section>
    );
  }
}

DueDate.propTypes = {
  dueDate: PropTypes.object.isRequired,
  t: PropTypes.func.isRequired,
  updateFieldData: PropTypes.func.isRequired,
};

// Export
export default _withFadeInAnimation(DueDate);
