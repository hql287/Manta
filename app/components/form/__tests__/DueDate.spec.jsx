// Libs
import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { SingleDatePicker } from 'react-dates';

// Component
import { DueDate } from '../DueDate.jsx';

// Mocks
const t = jest.fn();
const updateFieldData = jest.fn();
const dueDate = {
  required: true,
  selectedDate: {
    milliseconds: 0,
    seconds: 0,
    minutes: 0,
    hours: 12,
    date: 20,
    months: 10,
    years: 2017,
  },
};

describe('Renders correctly to the DOM', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(
      <DueDate t={t} dueDate={dueDate} updateFieldData={updateFieldData} />
    );
  });

  // TODO
  it('will not rerender if threre is no change in Props', () => {});

  // it('receives correct props', () => {
  //   expect(wrapper.prop('dueDate')).toEqual(dueDate);
  //   expect(wrapper.prop('dueDate')).not.toEqual({});
  //   expect(wrapper.prop('updateFieldData')).toEqual(updateFieldData);
  // });
  //
  // it('renders necessary element', () => {
  //   expect(wrapper.find('label')).toHaveLength(1);
  //   expect(wrapper.find('label')).not.toHaveLength(2);
  //   expect(wrapper.find(SingleDatePicker)).toHaveLength(1);
  //   expect(wrapper.find(SingleDatePicker)).not.toHaveLength(2);
  // });
  //
  // it('call clearDate when click clearDate Button', () => {
  //   // Setup
  //   const spy = jest.spyOn(DueDate.prototype, 'clearDate');
  //   const wrap = mount(
  //     <DueDate t={t} dueDate={dueDate} updateFieldData={updateFieldData} />
  //   );
  //   const clearDateBtn = wrap.find('.clearDateBtn');
  //   // Execute
  //   clearDateBtn.simulate('click');
  //   // Assertion
  //   expect(spy).toHaveBeenCalled();
  // });
  //
  // it('clearDate clears selectedDate', () => {
  //   // Setup
  //   const instance = wrapper.instance();
  //   const spy = jest.spyOn(instance, 'clearDate');
  //   // Execute
  //   instance.clearDate();
  //   // Assertion
  //   expect(spy).toHaveBeenCalled();
  //   expect(updateFieldData).toHaveBeenCalled();
  //   expect(updateFieldData).toHaveBeenCalledWith('dueDate', {
  //     selectedDate: null,
  //   });
  // });
  //
  // it('matches snapshot', () => {
  //   const tree = renderer
  //     .create(
  //       <DueDate t={t} dueDate={dueDate} updateFieldData={updateFieldData} />
  //     )
  //     .toJSON();
  //   expect(tree).toMatchSnapshot();
  // });
});
