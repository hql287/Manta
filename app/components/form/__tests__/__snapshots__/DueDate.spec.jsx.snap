// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly to the DOM matches snapshot 1`] = `
<div
  className="Section__SectionStyle-dgSKSv cwfAAu"
>
  <label
    className="itemLabel"
  />
  <div
    className="DueDate__DueDateContent-jGMeZA ddcSFi"
  >
    <select
      id="addDay"
      name="addDay"
      onChange={[Function]}
      placeholder={undefined}
      value={0}
    >
      <option
        value="0"
      />
      <option
        value="5"
      >
        +5 
      </option>
      <option
        value="10"
      >
        +10 
      </option>
      <option
        value="30"
      >
        +30 
      </option>
    </select>
    <div
      className="SingleDatePicker"
    >
      <div>
        <div
          className="SingleDatePickerInput"
        >
          <div
            className="DateInput DateInput--open-down"
          >
            <input
              aria-describedby="DateInput__screen-reader-message-invoice-duedate"
              aria-label="Date"
              autoComplete="off"
              className="DateInput__input needsclick"
              disabled={false}
              id="invoice-duedate"
              name="invoice-duedate"
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              placeholder="Date"
              readOnly={false}
              required={false}
              type="text"
              value="2017-11-20"
            />
            <p
              className="screen-reader-only"
              id="DateInput__screen-reader-message-invoice-duedate"
            >
              Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.
            </p>
            <div
              className="DateInput__display-text DateInput__display-text--has-input"
            >
              20/11/2017
            </div>
          </div>
        </div>
      </div>
    </div>
    <a
      className="clearDateBtn active"
      href="#"
      onClick={[Function]}
    >
      <i
        className="ion-close-circled"
      />
    </a>
  </div>
</div>
`;
